File: ch06/45/LastRemaining.cpp
class Solution {
public:
    int LastRemaining_Solution(int n, int m)
    {
        if (n < 1 || m < 1)
            return -1;

        list<int> numbers;
        for (int i = 0; i < n; i++)
            numbers.push_back(i);
        
        list<int>::iterator current = numbers.begin();
        while (numbers.size() > 1) {
            for (int j = 1; j < m; j ++) {
                current ++;
                if (current == numbers.end()) 
                    current = numbers.begin();
            }

            list<int>::iterator next = ++ current;
            if (next == numbers.end())
                next = numbers.begin();
            
            -- current;
            numbers.erase(current);
            current = next;
        }

        return (*current);
    }
};

File: ch06/45/LastRemaining.py
# -*- coding:utf-8 -*-
class Solution:
    def LastRemaining_Solution(self, n, m):
        # write code here
        if n < 1 or m < 1:
            return -1

        numbers = [x for x in range(n)]
        i = 0
        while len(numbers) > 1:
            for j in range(1, m):
                i += 1
                if i == len(numbers):
                    i = 0
            numbers.pop(i)
            if i == len(numbers):
                i = 0

        return numbers[0]

if __name__ == "__main__":
    s = Solution()
    print s.LastRemaining_Solution(5, 3)